node-webservice/react-client/src/Components/GameplayPage.jsx

```diff
@@ -32,40 +91,119 @@ class GameplayPage extends Component {
   renderGameplayUI() {
     // show different gameplay ui based on the gameplay phase
     if (this.state.gameplayPhase === 'buildrobot') {
-      return <RobotJSONObjectForm/>;
+      return (
+        <div className='gameplay-page'>
+          <h3>Playing game {this.props.match.params.gameid}</h3>
+          <Grid onCellClick={this.handleCellClicked} parts={this.state.parts}></Grid>
+        </div>
+      );
+      //return <RobotJSONObjectForm />;
     }
     if (this.state.gameplayPhase === 'controlrobot') {
-      return <h1>TODO: put robot controls here</h1>;
+      return (
+        <div className='gameplay-page'>
+          <h3>Playing game {this.props.match.params.gameid}</h3>
+          <Grid onCellClick={this.handleCellClicked} parts={this.state.parts}></Grid>
+        </div>
+      );
     }
     return <h1 className="error-message">Invalid gameplay phase: {this.state.gameplayPhase}</h1>
   }
 
   handleCellClicked(x, y) {
-    /*
-    if (x < 0 || x > 4) {
-      //error
+    //makes sure cell has valid coordinates
+    if (x < 0 || x > 4 || y < 0 || y > 4) {
+      throw new Error("Invalid x or y");
     }
-    if (this.state.selectedToolbarItem === 1) {
 
-    }*/
-    console.log("here: " + x + ", " + y);
+    if (this.state.gameplayPhase === "buildrobot") {
+      // Checks to see if there is a part in the clicked location. If there is,
+      // rotate the part and set 'partHere' to true.
+      var partHere = false;
+      var copy = [...this.state.parts];
+      copy.forEach(element => {
+        if (element.x === x && element.y === y) {
+          element.direction = this.rotate(element.direction, x, y);
+          partHere = true;
+        }
+      })
+      this.setState({ parts: copy })
+
+
+      if (!partHere) {
+        //Code to place selected part from the toolbar
+        var newPart = {
+          "type": "block",
+          "x": x,
+          "y": y,
+          "direction": "north",
+          "health": 1.0
+        }
+        copy.push(newPart);
+        this.setState({ parts: copy});
+      }
+    }
+
+
+  }
+  TwoDIndexOf(arr, target) {
+    var found = 0;
+    arr.forEach(element => {
+      if (element[0] === target[0] && element[1] === target[1]) {
+        found = 1;
+      }
+    })
+    return found;
+  }
+
+
+  rotate(current, x, y) {
+    const order = ["north", "west", "south", "east", "north", "west", "south", "east"]
+    var partLocations = [];
+    this.state.parts.forEach(element => {
+      partLocations.push([element.x, element.y])
+    })
+    var intial = order.indexOf(current);
+    if (intial === -1) {
+      throw new Error("invalid direction");
+    }
+    for (var a = intial + 1; a < intial + 4; a++) {
```

Comment:
This code seems to work well, but it is kinda weird to use an array with all the values twice. Not necessary to change right now, but to iterate over all elements an array from different starting points, you can make an index loop around by incrementing it like this `index = (index + 1) % arr.length`